File: oss/clawpipe/go-sdk/booster.go
package clawpipe

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// boosterRule is a deterministic transform that can skip the LLM.
type boosterRule struct {
	name    string
	test    func(string) bool
	resolve func(string) (string, error)
}

// Booster resolves prompts locally when possible.
type Booster struct{ rules []boosterRule }

// NewBooster creates a Booster with the six default rules.
func NewBooster() *Booster {
	b := &Booster{}
	b.rules = []boosterRule{
		b.jsonFormatRule(),
		b.mathRule(),
		b.dateRule(),
		b.unitConversionRule(),
		b.uuidRule(),
		b.base64Rule(),
	}
	return b
}

// TryResolve returns (answer, true) if a rule matched, else ("", false).
func (b *Booster) TryResolve(input string) (string, bool) {
	trimmed := strings.TrimSpace(input)
	for _, r := range b.rules {
		if r.test(trimmed) {
			if res, err := r.resolve(trimmed); err == nil {
				return res, true
			}
		}
	}
	return "", false
}

// RuleCount returns the number of registered rules.
func (b *Booster) RuleCount() int { return len(b.rules) }

// --- rules -----------------------------------------------------------

func (b *Booster) jsonFormatRule() boosterRule {
	return boosterRule{
		name: "json-format",
		test: func(s string) bool {
			l := strings.ToLower(s)
			return (strings.HasPrefix(l, "format this json") ||
				strings.HasPrefix(l, "pretty print")) && strings.Contains(s, "{")
		},
		resolve: func(s string) (string, error) {
			idx := strings.Index(s, "{")
			var v interface{}
			if err := json.Unmarshal([]byte(s[idx:]), &v); err != nil {
				return "", err
			}
			out, err := json.MarshalIndent(v, "", "  ")
			return string(out), err
		},
	}
}

var mathCmd = regexp.MustCompile(`(?i)^(?:calculate|compute|what is|evaluate|solve)\s+(.+)`)
var safeExpr = regexp.MustCompile(`^[\d\s+\-*/().,%^]+$`)

func (b *Booster) mathRule() boosterRule {
	return boosterRule{
		name: "math",
		test: func(s string) bool {
			m := mathCmd.FindStringSubmatch(s)
			return m != nil && safeExpr.MatchString(strings.TrimSpace(m[1]))
		},
		resolve: func(s string) (string, error) {
			m := mathCmd.FindStringSubmatch(s)
			expr := strings.ReplaceAll(strings.TrimSpace(m[1]), "^", "**")
			v, err := safeEvalMath(expr)
			if err != nil {
				return "", err
			}
			return formatNumber(v), nil
		},
	}
}

var datePats = []*regexp.Regexp{
	regexp.MustCompile(`(?i)what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)`),
	regexp.MustCompile(`(?i)(?:today|now|current date)`),
}

func (b *Booster) dateRule() boosterRule {
	return boosterRule{
		name: "date",
		test: func(s string) bool {
			if len(s) >= 60 {
				return false
			}
			for _, p := range datePats {
				if p.MatchString(s) {
					return true
				}
			}
			return false
		},
		resolve: func(string) (string, error) {
			return time.Now().UTC().Format(time.RFC3339), nil
		},
	}
}

var unitPat = regexp.MustCompile(`(?i)convert\s+([\d.]+)\s+(\w+)\s+to\s+(\w+)`)

type convFn func(float64) float64

func mul(f float64) convFn { return func(v float64) float64 { return v * f } }

var conversions = map[string]map[string]convFn{
	"km":    {"miles": mul(0.621371), "m": mul(1000), "ft": mul(3280.84)},
	"miles": {"km": mul(1.60934), "m": mul(1609.34), "ft": mul(5280)},
	"kg":    {"lbs": mul(2.20462), "g": mul(1000), "oz": mul(35.274)},
	"lbs":   {"kg": mul(0.453592), "g": mul(453.592), "oz": mul(16)},
	"c": {
		"f": func(v float64) float64 { return v*9/5 + 32 },
		"k": func(v float64) float64 { return v + 273.15 },
	},
	"f": {
		"c": func(v float64) float64 { return (v - 32) * 5 / 9 },
		"k": func(v float64) float64 { return (v-32)*5/9 + 273.15 },
	},
}

func (b *Booster) unitConversionRule() boosterRule {
	return boosterRule{
		name: "unit-conversion",
		test: func(s string) bool { return unitPat.MatchString(s) },
		resolve: func(s string) (string, error) {
			m := unitPat.FindStringSubmatch(s)
			val, _ := strconv.ParseFloat(m[1], 64)
			from, to := strings.ToLower(m[2]), strings.ToLower(m[3])
			fn, ok := conversions[from][to]
			if !ok {
				return "", fmt.Errorf("unknown conversion %s->%s", from, to)
			}
			res := fn(val)
			rounded := math.Round(res*10000) / 10000
			return fmt.Sprintf("%v %s = %v %s", val, from, rounded, to), nil
		},
	}
}

var uuidPat = regexp.MustCompile(`(?i)generate\s+(?:a\s+)?uuid`)

func (b *Booster) uuidRule() boosterRule {
	return boosterRule{
		name: "uuid",
		test: func(s string) bool { return uuidPat.MatchString(s) },
		resolve: func(string) (string, error) { return generateUUID(), nil },
	}
}

var b64Enc = regexp.MustCompile(`(?i)base64\s+encode\s+(.+)`)
var b64Dec = regexp.MustCompile(`(?i)base64\s+decode\s+(.+)`)

func (b *Booster) base64Rule() boosterRule {
	return boosterRule{
		name: "base64",
		test: func(s string) bool { return b64Enc.MatchString(s) || b64Dec.MatchString(s) },
		resolve: func(s string) (string, error) {
			if m := b64Enc.FindStringSubmatch(s); m != nil {
				return base64.StdEncoding.EncodeToString(
					[]byte(strings.TrimSpace(m[1]))), nil
			}
			if m := b64Dec.FindStringSubmatch(s); m != nil {
				d, err := base64.StdEncoding.DecodeString(strings.TrimSpace(m[1]))
				return string(d), err
			}
			return "", fmt.Errorf("no match")
		},
	}
}

File: oss/clawpipe/python-sdk/clawpipe/booster.py
"""Agent Booster -- deterministic transforms that skip LLM calls.

Resolves prompts locally when the answer can be computed without AI:
JSON formatting, math, date/time, unit conversion, UUID, base64.
"""

from __future__ import annotations

import base64
import json
import re
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Callable, Optional

from .math_eval import safe_eval_math


@dataclass
class BoosterRule:
    """A single booster rule with a test predicate and resolver."""

    name: str
    test: Callable[[str], bool]
    resolve: Callable[[str], str]


class Booster:
    """Try to resolve prompts without calling an LLM."""

    def __init__(self) -> None:
        self._rules: list[BoosterRule] = []
        self._register_defaults()

    def try_resolve(self, input_text: str) -> Optional[str]:
        """Try each rule. Return resolved text or None."""
        trimmed = input_text.strip()
        for rule in self._rules:
            if rule.test(trimmed):
                try:
                    return rule.resolve(trimmed)
                except Exception:
                    continue
        return None

    def add_rule(self, rule: BoosterRule) -> None:
        """Register a custom booster rule."""
        self._rules.append(rule)

    @property
    def rule_count(self) -> int:
        return len(self._rules)

    def _register_defaults(self) -> None:
        self._rules.extend([
            self._json_format_rule(),
            self._math_rule(),
            self._date_rule(),
            self._unit_conversion_rule(),
            self._uuid_rule(),
            self._base64_rule(),
        ])

    def _json_format_rule(self) -> BoosterRule:
        def test(inp: str) -> bool:
            lower = inp.lower()
            return (
                lower.startswith("format this json")
                or lower.startswith("pretty print")
            ) and "{" in inp

        def resolve(inp: str) -> str:
            idx = inp.index("{")
            return json.dumps(json.loads(inp[idx:]), indent=2)

        return BoosterRule(name="json-format", test=test, resolve=resolve)

    def _math_rule(self) -> BoosterRule:
        math_pat = re.compile(
            r"^(?:calculate|compute|what is|evaluate|solve)\s+(.+)", re.I
        )
        safe_expr_pat = re.compile(r"^[\d\s+\-*/().,%^]+$")

        def test(inp: str) -> bool:
            m = math_pat.match(inp)
            if not m:
                return False
            return bool(safe_expr_pat.match(m.group(1).strip()))

        def resolve(inp: str) -> str:
            m = math_pat.match(inp)
            assert m is not None
            expr = m.group(1).strip().replace("^", "**")
            result = safe_eval_math(expr)
            if result == int(result):
                return str(int(result))
            return str(result)

        return BoosterRule(name="math", test=test, resolve=resolve)

    def _date_rule(self) -> BoosterRule:
        patterns = [
            re.compile(r"what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)", re.I),
            re.compile(r"(?:today|now|current date)", re.I),
        ]

        def test(inp: str) -> bool:
            return any(p.search(inp) for p in patterns) and len(inp) < 60

        def resolve(_inp: str) -> str:
            return datetime.now(timezone.utc).isoformat()

        return BoosterRule(name="date", test=test, resolve=resolve)

    def _unit_conversion_rule(self) -> BoosterRule:
        pattern = re.compile(r"convert\s+([\d.]+)\s+(\w+)\s+to\s+(\w+)", re.I)
        conversions: dict[str, dict[str, float | Callable[[float], float]]] = {
            "km": {"miles": 0.621371, "m": 1000, "ft": 3280.84},
            "miles": {"km": 1.60934, "m": 1609.34, "ft": 5280},
            "kg": {"lbs": 2.20462, "g": 1000, "oz": 35.274},
            "lbs": {"kg": 0.453592, "g": 453.592, "oz": 16},
            "c": {
                "f": lambda v: v * 9 / 5 + 32,
                "k": lambda v: v + 273.15,
            },
            "f": {
                "c": lambda v: (v - 32) * 5 / 9,
                "k": lambda v: (v - 32) * 5 / 9 + 273.15,
            },
        }

        def test(inp: str) -> bool:
            return bool(pattern.search(inp))

        def resolve(inp: str) -> str:
            m = pattern.search(inp)
            assert m is not None
            value = float(m.group(1))
            from_unit = m.group(2).lower()
            to_unit = m.group(3).lower()
            conv = conversions.get(from_unit, {}).get(to_unit)
            if conv is None:
                raise ValueError("Unknown conversion")
            result = conv(value) if callable(conv) else value * conv
            rounded = round(result, 4)
            # Strip trailing zeros like JS Number()
            formatted = f"{rounded:g}" if rounded != int(rounded) else str(int(rounded))
            return f"{value:g} {from_unit} = {formatted} {to_unit}"

        return BoosterRule(name="unit-conversion", test=test, resolve=resolve)

    def _uuid_rule(self) -> BoosterRule:
        pat = re.compile(r"generate\s+(?:a\s+)?uuid", re.I)

        def test(inp: str) -> bool:
            return bool(pat.search(inp))

        def resolve(_inp: str) -> str:
            return str(uuid.uuid4())

        return BoosterRule(name="uuid", test=test, resolve=resolve)

    def _base64_rule(self) -> BoosterRule:
        encode_pat = re.compile(r"base64\s+encode\s+(.+)", re.I)
        decode_pat = re.compile(r"base64\s+decode\s+(.+)", re.I)

        def test(inp: str) -> bool:
            return bool(encode_pat.search(inp) or decode_pat.search(inp))

        def resolve(inp: str) -> str:
            enc = encode_pat.search(inp)
            if enc:
                return base64.b64encode(enc.group(1).strip().encode()).decode()
            dec = decode_pat.search(inp)
            if dec:
                return base64.b64decode(dec.group(1).strip()).decode("utf-8")
            raise ValueError("No match")

        return BoosterRule(name="base64", test=test, resolve=resolve)
